### DISCUSSION TITLE
[Question] Meteor/Apollo Subscriptions

### DISCUSSION THREAD

**Discussion by ajaypillay** [enisdenjo/graphql-ws#199] [Q&A] [answered]
Hi, I'm running a Meteor/Apollo/React stack and I was trying to get Graphql Subscriptions working and I came across this package. However, there isn't an example of implementation using Meteor/Apollo.

Would anyone be able to advise on how to proceed with this?

Currently on my server I have:
```js
const schema = makeExecutableSchema({
    typeDefs,
    resolvers,
})
const context = async ({req}) => ({
    user: await getUser(req.headers.authorization)
})
const server = new ApolloServer({
    schema,
    context
})
server.applyMiddleware({
    app: WebApp.connectHandlers,
    path: "/graphql",
})
WebApp.connectHandlers.use("/graphql", (req, res) => {
    if (req.method === "GET") {
        res.end()
    }
})
```
And on my client I have:
```js
const httpLink = new HttpLink({
    uri: "http://XXX.XXX.XX.XX:3000/graphql"
})
const authLink = new ApolloLink((operation, forward) => {
    const token = Accounts._storedLoginToken()
    operation.setContext(() => ({
        headers: {
            authorization: token
        }
    }))
    return forward(operation)
})
const client = new ApolloClient({
    link: from([authLink, httpLink]),
    cache: new InMemoryCache()
})
const ApolloApp = () => (
    <BrowserRouter>
        <ApolloProvider client={client}>
            <App client={client} />
        </ApolloProvider>
    </BrowserRouter>
)
Meteor.startup(() => {
    render(<ApolloApp />, document.getElementById("app"))
})
```
How should I go about configuring this to allow for subscriptions? Meteor does have the [swydo:ddp-apollo](https://github.com/Swydo/ddp-apollo) package but I've not been able to set that up correctly and one of the drawbacks is that Graphiql will no longer be available. I can't depend on Apollo Dev Tools right now either because it doesn't open up Graphiql either regardless of server configuration (already filed a bug for this, it just shows up as blank).

If there's some way I can use this package to get subscriptions working, that would be great. Appreciate any help, thank you.

**Comment by enisdenjo** [comment]
Hey there! Sadly I have no experience with Apollo whatsoever, and very little with Meteor (years ago I've used it).

Briefly looking at the mentioned `ddp-apollo` library, seems like it uses [a custom WebSocket protocol for the ApolloLink](https://github.com/Swydo/ddp-apollo/blob/838abfaa17cf6b1291ffbf71035b694d2bbb235b/packages/apollo-link-ddp/src/client/apollo-link-ddp.js#L68-L122). For `graphql-ws` to work, both the client and the server must implement [the GraphQL over WebSocket Protocol](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md). There is, however, a ["Client usage with Apollo" recipe](https://github.com/enisdenjo/graphql-ws#apollo-client) showcasing how to create a `graphql-ws` compatible ApolloLink.

Regarding the server, [Meteor exports the `webapp` package](https://docs.meteor.com/packages/webapp.html) that lets you hook up your own server logic. Dont know how exactly, but I am very certain that you can provide your own WebSocket server implementation (use the abundance of [ws server recipes](https://github.com/enisdenjo/graphql-ws#recipes) to help you start).

**Comment by ajaypillay** [comment]
Hi, sorry for the late response as I had to work on other aspects of the app I'm building.

I will be working on this shortly and will look into your suggestions and will update here once I have made any progress. Thanks a lot for the explanation and links.

**Reply by enisdenjo** [reply]
@ajaypillay converted this to a discussion. Fits better here.

**Reply by ajaypillay** [reply]
Okay got it! Still haven't had the chance to get around to this but I'll update once I do :-)

**Reply by ajaypillay** [reply]
Hi @enisdenjo, so I've finally gotten the chance to work on this. I'd like to link you to [this thread I posted](https://community.apollographql.com/t/graphql-subscriptions-working-but-i-dont-quite-understand-why-how/574) on how I got subscriptions working with Meteor/Apollo. Note that I did not have to use this package to get subscriptions working.

I am curious though - am I doing something wrong that this package can solve/help with? Disregarding Meteor, it seems that anyone can get subscriptions working with the same code I have using the original `ApolloClient` and `ApolloServer` with no middleware. I understand that this package has many other integrations, but if one were to be using ApolloClient and ApolloServer, does this package fill any gaps which the original implementations miss?

**Reply by enisdenjo** [reply]
Both [`ApolloServer`](https://github.com/apollographql/apollo-server/blob/8d1179ed007b90c56382826499011af0dde16439/packages/apollo-server-core/src/ApolloServer.ts#L899-L1007) and [Apollo's `WebSocketLink`](https://github.com/apollographql/apollo-client/blob/cbcf951256b22553bdb065dfa0d32c0a4ca804d3/src/link/ws/index.ts) use the legacy [`subscriptions-transport-ws`](https://github.com/apollographql/subscriptions-transport-ws) by default.

So, you're actually not even using this library. 😄

**Reply by ajaypillay** [reply]
I'm not sure if you're the right person to ask - but do you know why these libraries are relying on a legacy/unmaintained package when yours exists? It seems quite odd that this is the case.

**Reply by enisdenjo** [reply]
Apollo is huge. My bet is that they simply have no time at their hands to do the switch. Guess they're waiting for PRs (and I am not filing one because _I_ dont have the time haha). I did take a brief look in Apollo Server, doing the switch is not that straight forward.

However, they do plan on dropping the `subscriptions-transport-ws` integration starting Apollo Server v3. You can find the following on the roadmap (https://github.com/apollographql/apollo-server/blob/main/ROADMAP.md):
> Removal of subscription-transport-ws in favor of a built-in solution that takes full advantage of the main request pipeline.

This means that integrating `graphql-ws` might not be an option either, they might just want a bare minimum server with external plugins adding the necessary integrations.

There is also an issue for the server about documenting the migration to `graphql-ws` or preserving `subscriptions-transport-ws` (https://github.com/apollographql/apollo-server/issues/4960).

Doing so with the server, I guess, the client will follow suite and either support both, or drop the integration and rely on a plugin system.

**Reply by ajaypillay** [reply]
Thank you so much for the explanation. I completely recognize the complexity of Apollo and don't have the knowledge/experience to understand it fully. It's great that it's known and being worked on. So I've gone ahead and migrated my code over to use this package.

Client code:
```js
import React from "react"
import { render } from "react-dom"
import { BrowserRouter } from "react-router-dom"
import { ApolloClient, ApolloProvider, HttpLink, split, InMemoryCache } from "@apollo/client"
import { ApolloLink, Observable} from '@apollo/client/core'
import { offsetLimitPagination, getMainDefinition } from "@apollo/client/utilities"
import { ApolloLink as ApolloLink2, from } from "apollo-link"
import { print, GraphQLError } from "graphql"
import { createClient } from "graphql-ws"
import App from "./some_path/App"
  
class WebSocketLink extends ApolloLink {
    constructor(options) {
        super()
        this.client = createClient(options)
    }

    request(operation) {
        return new Observable((sink) => {
            return this.client.subscribe(
            { ...operation, query: print(operation.query) },
            {
            next: sink.next.bind(sink),
            complete: sink.complete.bind(sink),
            error: (err) => {
                if (err instanceof Error) return sink.error(err)

                if (err instanceof CloseEvent) {
                    return sink.error(
                        // reason will be available on clean closes
                        new Error(
                        `Socket closed with event ${err.code} ${err.reason || ''}`
                        )
                    )
                }

                return sink.error(new GraphQLError({ message: message }))
            }
            }
            )
      })
    }
}

const httpLink = new HttpLink({ uri: "http://XXX.XXX.XX.XX:4000/graphql" })

const authLink = new ApolloLink2((operation, forward) => {
    const token = Accounts._storedLoginToken()
    operation.setContext(() => ({
        headers: {
            authorization: token
        }
    }))
    return forward(operation)
})

const wsLink = new WebSocketLink({ url: "ws://XXX.XXX.XX.XX:4000/subscriptions" })

const splitLink = split(
    ({ query }) => {
      const definition = getMainDefinition(query)
      return (
        definition.kind === 'OperationDefinition' &&
        definition.operation === 'subscription'
      )
    },
    wsLink,
    httpLink
)

const cache = new InMemoryCache({
    typePolicies: {
        Query: {
            fields: {
                getFeed: offsetLimitPagination()
            }
        }
    }
})

const client = new ApolloClient({
    link: from([authLink, splitLink]),
    cache
})

const ApolloApp = () => (
    <BrowserRouter>
        <ApolloProvider client={client}>
            <App client={client} />
        </ApolloProvider>
    </BrowserRouter>
)

Meteor.startup(() => {
    render(<ApolloApp />, document.getElementById("app"))
})

// The following line is to disconnect Meteor's DDP/WebSocket connection
// as it is not needed in production. Can be commented out in development
// to allow for Hot Code Push convenience.
Meteor.disconnect()
```
And here is my server code:
```js
import express from "express"
import { ApolloServer } from "apollo-server-express"
import ws from "ws"
import { useServer } from "graphql-ws/lib/use/ws"

const typeDefs = [/* Schemas */]
const resolvers = merge(/* Resolvers */)

const schema = makeExecutableSchema({ typeDefs, resolvers })

const app = express()

const context = async ({req}) => ({
    user: await getUser(req?.headers.authorization)
})

const apolloServer = new ApolloServer({ schema, context })

apolloServer.applyMiddleware({ app })

const server = app.listen(4000, () => {
    const wsServer = new ws.Server({
        server,
        path: '/subscriptions'
    })
    useServer({ schema }, wsServer)
})
```
Some notes:
1) For the client code, I modified it as it was in TypeScript (my entire project is in JS not TS, and I'm not experienced with TS, though I will be learning it soon).
2) This is basically an Apollo/Express setup. Not too sure what Meteor's role is over here apart from allowing me to pump in the user token into `context`, and of course for startup.

**Reply by enisdenjo** [reply]
I am really not experienced with Apollo or Meteor - to me this looks just fine.

Only one thing I'd like to bring up: you're not using the `context` with `graphql-ws` on the server. Maybe you dont need it, but I guess you do. 😄 
```diff
- useServer({ schema }, wsServer)
+ useServer({ schema, context }, wsServer) 
```

You'll also have to modify the `context` function because the `req` is not present in the first argument. You'll need to authenticate the user differently for WS. (There are recipes to help you with this)

Good luck with TS, IMHO it is a must have in anyones tool belt. 😉

**Reply by ajaypillay** [reply]
Yep no worries! Thank you for raising that, I figured out how to do this.

Client:
```js
const wsLink = new WebSocketLink({
    url: "ws://XXX.XXX.XX.XX:4000/subscriptions",
    connectionParams: () => ({ authorization: Accounts._storedLoginToken() })
})
```
Server:
```js

const wsContext = async ({ connectionParams }) => ({
    user: await getUser(connectionParams.authorization)
})
...
useServer({ schema, context: wsContext }, wsServer)
```

Thank you so much again for the help.

**Reply by enisdenjo** [reply]
Great stuff! You're welcome. 😁

**Reply by enisdenjo** [reply]
Would be great if you could post your final solution as a suggested answer below and mark it as the answer for others to find.

**Reply by ajaypillay** [reply]
Will do! EDIT: Moved the code from this reply into an answer.

**Comment by ajaypillay** [comment, marked as answer]
Here's the final complete source for my client and server. For more insights, check out [this thread](https://community.apollographql.com/t/graphql-subscriptions-working-but-i-dont-quite-understand-why-how/574):

Client:
```js

import React from "react"
import { render } from "react-dom"
import { BrowserRouter } from "react-router-dom"
import { ApolloClient, ApolloProvider, HttpLink, split, InMemoryCache } from "@apollo/client"
import { ApolloLink, Observable} from '@apollo/client/core'
import { offsetLimitPagination, getMainDefinition } from "@apollo/client/utilities"
import { ApolloLink as ApolloLink2, from } from "apollo-link"
import { print, GraphQLError } from "graphql"
import { createClient } from "graphql-ws"
import App from "../../ui/App"
  
class WebSocketLink extends ApolloLink {
    constructor(options) {
        super()
        this.client = createClient(options)
    }

    request(operation) {
        return new Observable((sink) => {
            return this.client.subscribe(
            { ...operation, query: print(operation.query) },
            {
            next: sink.next.bind(sink),
            complete: sink.complete.bind(sink),
            error: (err) => {
                if (err instanceof Error) return sink.error(err)

                if (err instanceof CloseEvent) {
                    return sink.error(
                        // reason will be available on clean closes
                        new Error(
                        `Socket closed with event ${err.code} ${err.reason || ''}`
                        )
                    )
                }

                return sink.error(new GraphQLError({ message: message }))
            }
            }
            )
      })
    }
}

const httpLink = new HttpLink({ uri: "http://XXX.XXX.XX.XX:4000/graphql" })

const authLink = new ApolloLink2((operation, forward) => {
    const token = Accounts._storedLoginToken()
    operation.setContext(() => ({
        headers: {
            authorization: token
        }
    }))
    return forward(operation)
})

const wsLink = new WebSocketLink({
    url: "ws://XXX.XXX.XX.XX:4000/subscriptions",
    connectionParams: () => ({ authorization: Accounts._storedLoginToken() })
})

const splitLink = split(
    ({ query }) => {
      const definition = getMainDefinition(query)
      return (
        definition.kind === 'OperationDefinition' &&
        definition.operation === 'subscription'
      )
    },
    wsLink,
    httpLink
)
// You can just use a normal InMemoryCache with no options, I just left
// my implementation in
const cache = new InMemoryCache({
    typePolicies: {
        Query: {
            fields: {
                getFeed: offsetLimitPagination()
            }
        }
    }
})

const client = new ApolloClient({
    link: from([authLink, splitLink]),
    cache
})

const ApolloApp = () => (
    <BrowserRouter>
        <ApolloProvider client={client}>
            <App client={client} />
        </ApolloProvider>
    </BrowserRouter>
)

Meteor.startup(() => {
    render(<ApolloApp />, document.getElementById("app"))
})
/*
The following line is to disconnect Meteor's DDP/WebSocket connection
as it is not needed in production. Can be commented out in development
to allow for Hot Code Push convenience.
*/
//Meteor.disconnect()
```
Server:
```js

import express from "express"
import { ApolloServer } from "apollo-server-express"
import ws from "ws"
import { useServer } from "graphql-ws/lib/use/ws"

const typeDefs = [/* Schemas */]
const resolvers = merge(/* Resolvers */)

const schema = makeExecutableSchema({ typeDefs, resolvers })

const app = express()

const context = async ({req}) => ({
     user: await getUser(req.headers.authorization) 
})

const wsContext = async ({ connectionParams }) => ({
    user: await getUser(connectionParams.authorization)
})

const apolloServer = new ApolloServer({ schema, context })

apolloServer.applyMiddleware({ app })

const server = app.listen(4000, () => {
    const wsServer = new ws.Server({
        server,
        path: '/subscriptions'
    })

    useServer({ schema, context: wsContext }, wsServer)
})
```
